log/wrapper.go

```diff
@@ -127,7 +129,28 @@ func (w *Wrapper) UnmarshalText(text []byte) error {
 	// Special handling for "zap:level" case
 	const zapLevelPrefix = "zap:"
 	if strings.HasPrefix(s, zapLevelPrefix) {
-		*w = ZapWrapper(Level(strings.ToLower(s[len(zapLevelPrefix):])))
+		errMalform := fmt.Errorf("log.Wrapper.UnmarshalText: malformed input: %q", s)
+		split := strings.Split(s, ":")
+		if len(split) > 3 {
+			return errMalform
+		}
+		var pairs map[string]interface{}
+		if len(split) > 2 {
+			kvs := strings.Split(split[2], ",")
+			pairs = make(map[string]interface{}, len(kvs))
+			for _, kv := range kvs {
+				kv = strings.TrimSpace(kv)
+				index := strings.Index(kv, "=")
+				if index < 0 {
+					return errMalform
+				}
+				pairs[kv[:index]] = kv[index+1:]
+			}
+		}
+		*w = ZapWrapper(ZapWrapperArgs{
+			Level:   Level(strings.ToLower(split[1])),
+			KVPairs: pairs,
+		})
```

Comment:
It's not saving you anything by "precomputing" this -- you'll only ever compute it once.  Now, however, you also compute this in the happy path.

This also doesn't say _how_ it's malformed.  Each return path should have a unique error message to make it clear both to the reader and someone who is trying to follow the error through the code what is wrong.

```suggestion
		split := strings.Split(s, ":")
		if len(split) > 3 {
			return fmt.Errorf("log.Wrapper.UnmarshalText: malformed input: too many arguments in %q", s)
		}
		var pairs map[string]interface{}
		if len(split) > 2 {
			kvs := strings.Split(split[2], ",")
			pairs = make(map[string]interface{}, len(kvs))
			for _, kv := range kvs {
				kv = strings.TrimSpace(kv)
				index := strings.Index(kv, "=")
				if index < 0 {
					return fmt.Errorf("log.Wrapper.UnmarshalText: malformed input: keyval pair %q has no = in %q", kv, s)
				}
				pairs[kv[:index]] = kv[index+1:]
			}
		}
		*w = ZapWrapper(ZapWrapperArgs{
			Level:   Level(strings.ToLower(split[1])),
			KVPairs: pairs,
		})
```

Reply:
I'm not sure what your concern is in:

>It's not saving you anything by "precomputing" this -- you'll only ever compute it once. Now, however, you also compute this in the happy path.

but I addressed your other concerns.

Reply:
You fixed it.  I was saying that you were doing the `Errorf` even in non-error cases